### DISCUSSION TITLE
New structure for database

### DISCUSSION THREAD

**Discussion by rebio** [lensfun/lensfun#1616] [Ideas]
Hello there!

I was going through the structure of the camera and lens database and even I'm sure that there is a good reason why it's the way it is, I that that it might be a good idea to resort it a bit and give it a more defined structure.
I have to admit right from the start that I only looked into the database because I was searching for a particular lens and I found it to be very messy. Yes, with a structure, but messy and everything is all over the place.

If you are a (lead) programer of this project and can only shake your head about this suggestion and think that I have no idea and must have written this from a mental asylum or otherwise I would not be able to come up with this level of stupidity, you are at least right about the first one.
If that's the case, feel free to tell me so and may you can invest some time to tell me why that is. If every link to a file is hardcoded in lensfun and you would have to change ALL links it would be for sure tedious, but far from impossible.

**Why "all over the place":**
- Cameras and lenses are mixed within a file
- Drones are sorted in "actioncams.xml" but some smartphones are sorted in "misc.xml". Other smartphones are sorted somewhere else (eg. the Galaxy Note 8 sorted in "compact-samsung.xml"). Why is that sorting? Sensor size? Quality? So that there is not one giant file?
- Sorted by _system class_ (= slr, mil) seems to be very weird. I could use eg. the Canon EF-S 60 mm Macro on a Canon EOS R5. So one setup would more than one system class (not to mention that you could attach eg. the R5 to a home telescope for astrophotography. Where would this be sorted in?)

**What I noticed so far for the database:**
- Sorted by _system class_ and brand (System class = slr, mil, actioncams)
- Camera and lenses are not sorted within a file
- Uncommon cameras (eg. in smartphones and drones) are not sorted with a pattern

**My suggested changes:**
- Sort by type (lens, camera, etc.)
- Sort by brand (Canon, Nikon, etc.)
- Sort by system class (slr, mil, etc. - if needed/wanted)
- One file per sorting topic (so no drones in _actioncams.xml_ and no smartphones in _misc.xml_)

**I would think of something like this (some names as example):**
- canon-lens[-slr] (all [only the slr] canon lenses would be sorted here; name could also be slr_lens)
- canon-camera[-slr] (all [only the slr] canon cameras would be sorted here; name could also be slr_camera)
- canon-compact_camera (all canon compact cameras would be sorted here)
- gopro-action_camera (all GoPro action cameras would be sorted here; named as action|SPACE|camera, because of it's article on Wikipedia)
- samsung-portable_device (all Samsung smartphones, tablets, etc. would be sorted here)
- dji-drone (all DJI drones would be sorted here, this would not include other products like their "Action" or "Ronin" products)

**Naming scheme:**
- brand name first, then the rest (if you know anything about a device you want to add/take a look at the specs it's the brand name)
- singular, not plural (so _camera_, not _cameras_)
- sorted by brand, not manufacturer (so _Google NEXUS 5_, not _LG NEXUS 5_)
- placeholder for every lens (if there is no correction data about a lens there should be an entry. If it's not possible to create an "empty" entry with no info, the entry could be added as comment
- lenses and cameras are only separated if they are interchangable. If they are in one unit (eg. portable devices, action cameras, compact cameras, etc.) and are not interchangeable they should be in one file

**Mics, Q&A, etc.:**
- That would be a lot of new files with few entries!
Yes, but there already are these files (eg. compact-kodak.xml, compact-samsung.xml, compact-leica.xml, compact-sigma.xml, contax.xml) and they are not sorted in misc.xml
- Many lenses have an item on Wikidata. Maybe we could link that. Either with a tag or a comment. Usage: Unclear right now, but I did a lot of things like this in the past that helped a lot with structure as well as finding stuff and sorting stuff later on. And if we would sort the whole database, this would probably rather trivial to add.
- A problem I just though about: What if you have a eg. Tamron lens for Canon and Nikon cameras. This would be solved by the current setup (eg. Tamron lens with Canon EF mount in _slr-Canon_ and _slr-Nikon_ for the F mount). However, this could also be achieved by adding two entries to the eg. tamron-lens.xml with two different mounts
- Does lensfun expect the lens and camera to be in one file? (I have no idea)
- Is there any reason why one lens <model> tag is _Canon RF 24-105mm F4-7.1 IS STM_ while the other one is  just _RF24-105mm F4 L IS USM_. Is this set by the manufacturer or the internal naming scheme. If it's set by lensfun I would suggest to stick to one standard (which is just how the manufacturer does it)

Please let me know what you think of my idea!
I could start the change in a branch in a subfolder of the database so they could coexist side-by-side for a migration process, but I would like to hear you comments first.

**Comment by rebio** [comment]
So that there is a better understanding of what I have in mind I forked the repo and started working in a branch: https://github.com/rebio/lensfun/tree/brand-based-database-structure

**Reply by uchrisu** [reply]
On My Opinion its a great idea to give the database a better (more consistent) structure.

I doubt that reorganizing in a forked repo is the best idea since there is much progress in the database in the last days and it would be rather hard to keep an overview what has moved and what was added. 

`A problem I just though about: What if you have a eg. Tamron lens for Canon and Nikon cameras. This would be solved by the current setup (eg. Tamron lens with Canon EF mount in slr-Canon and slr-Nikon for the F mount). However, this could also be achieved by adding two entries to the eg. tamron-lens.xml with two different mounts`

I think in most cases it should have the same correction data. So most times, one entry should be enough. Maybe we have to change code that this entry is recognized twice then for both mounts? I'm not sure at the moment if this can easily be done. 

`Is there any reason why one lens tag is Canon RF 24-105mm F4-7.1 IS STM while the other one is just RF24-105mm F4 L IS USM. Is this set by the manufacturer or the internal naming scheme. If it's set by lensfun I would suggest to stick to one standard (which is just how the manufacturer does it)`

It is the name that exiv2 returns so that the lens can automatically be found. Here the same idea as above: We have many issues that this name can change on different mounts/adaptors/firmwares/... Maybe it is possible to have multiple tags here. Probably the code needs some updates for this.

**Reply by rebio** [reply]
You are right. It would be best to have the new database on the same level as the old one. So I can create a pull request if you want.
Even though it wouldn't be _that_ hard to include the later changes to the new database. If you have a few lenses per commit, they can be done rather quickly. Transfering the whole database will take much longer than adding a few lines that can easily be copied from the RAW view.

The question is who would do the changes though. Even I could do some more migrational work, I'm not sure if I could even commit to a branch in this repo if I don't own some kind of writing right. I would have no problems with that and I could limit my work to the branch, but since I know that it's much about trust in FOSS projects, I don't want to force myself into the game.

**Comment by rebio** [comment]
A new feature for the database: Add a _score_ or quality level to a calibration.

There are some calibration entries that have detailed information (eg. Canon EF-M 18-55mm f/3.5-5.6 IS STM) and some that have very little information (eg. Canon EF-M 28mm f/3.5 Macro IS STM). The scoring system would clearly show on the first look if the calibration entry could or maybe even should be improved.

**Reply by rebio** [reply]
@uchrisu I created a pull request for the fork branch on #1649

**Reply by gnl21** [reply]
> A new feature for the database: Add a score or quality level to a calibration.

This seems like a very useful feature for the dashboard showing lens support, but it doesn't seem to belong in the database to me. Are you thinking of using anything that can't easily be computed by a script parsing the database entries?

**Comment by Sturmflut** [comment]
I think the most distinguishing factors are the names of the brand and the mount. Many brands have enough products in the database to even split things further down into cameras and lenses, e.g. `nikon-f-camera.xml`, `canon-ef-lens.xml`. I don't see the "class" a helpful since it is usually not part of any naming scheme, in contrast I see it as problematic - I often had to look up the class when calibrating uploaded images since I am not very well-versed with anything other than Nikon and Sony.

Manufacturers that don't have many products in the database can be handled in a single file, e.g. `zeiss.xml`. `misc.xml` doesn't contain so many manufacturers that this would lead to a very cluttered database directory.

If there is no mount, but there are enough devices to make e.g. a manufacturer file like `samsung.xml` too cluttered, there can still e.g. be a `samsung-compact.xml` or `samsung-smartphone.xml`.

Mounts should be handled in a single file, `mounts.xml`, because so many of them reference each other.

Cameras should be sorted alphabetically since there is no other information available than the name.

Sorting within a file should be by minimal focal length (ascending), then by minimum aperture (ascending) and then by crop factor (ascending). This makes it easy to find a lens/the correct place and generally speaking sorts entries in the same region by quality.

I am not a huge friend of marking the same lens with multiple `<mount>` entries. I know many say the same lens constructed for multiple mounts is optically equivalent, but that doesn't feel true and is not what I've seen when calibrating images. There are small differences, as is to be expected. So I'm more in favor of only adding multiple `<mount>` entries if optical properties have been proven to be identical. Inconsistent naming for the same lens built for multiple mounts is something that has to be fixed either in exiv2 - or can't be fixed because the manufacturers themselves use different names.

I don't think we should put too much thought into converters/adapters. Converters that don't interfere with optical properties usually also don't interfere with the EXIF data, e.g. the Nikon FTZ and Canon EOS-R EF adapter don't. Converters that interfere with things make life hard anyways.

As to lens naming: That's different for each manufacturer. Some put their own name into their `Lens` fields, some don't. I agree that things should be consistent, and there should be a logic that tries to find the matching entry. For example: The database entry for the Nikon AF-S Nikkor 300mm f/4E PF ED VR should always be

```
<maker>Nikon</maker>
<model>AF-S Nikkor 300mm f/4E PF ED VR</model>
```

and the lookup logic should be intelligent enough to also try maker+model instead of just model.

**Comment by rebio** [comment]
@Sturmflut Numbers are the paragraphs:
Add 1)
The idea was that the database is structured so simple that a pre-school child would find an entry. If there is a difference in performance between many tiny files vs. few large files then performance should dictate how to handle the database. If not, I would collect all eg. camera entries in one big file. That way you do not have to go through multiple files when you search or a camera. We could split the lens entries by lens mount, but eg. for lenses with Canon FL lens mount it would be a file with only two entries. You could also make a canon-misc.xml file, but then you would have the same mess again and the same question arise _Where do I sort lens X_. And you would have to do the same for eg. Tamron or sigma. You would have a bunch of files that may contain duplicate entries (because of 7) aka same corrective data, different lens mount)
But not entirely oppose of it. Since I split by lens mount this can be done rather fast.

Add 2)
misc.xml is may needed, but I see the potential problem that you again have to search around where to put a new entry. Eg. a new Apple smartphone. Currently sorted in misc even there are quite a few models. If more Apple devices are added, at what number should they be moved to their separate file?

Add 3)
Done so in the pull request. There is _samsung-compact_camera.xml_ (for compact cameras) and _samsung-portable_device.xml_ (for smartphones, tables, etc.)

Add 4)
Thats a good idea. I just didn't know if the app expects to have a list of lens mounts in every file (so you would have to list all lens mounts in a camera as well as a lens file. Thats why I thought it could be the case that they are split into _mil-_ and _slr-_). If multiple the database files are treated like a single big one: Yes, all lens mounts should be collected in a single file.

Add 5)
Done so in the pull request.

Add 6)
Done so in the pull request except the _minimum aperture_. I guess you meant maximum aperture, right? My sorting is eg.
- EF 24mm f/2.8
- TS-E 24mm f/3.5 (<- name is ignored, only sorted by min focal length)
- EF 24mm f/4.0
- EF 24-70mm f/2.8
- EF 24-70mm f/2.8 II
- EF 24-70mm f/4.0 IS USM (crop factor 1.0)
- EF 24-70mm f/4.0 IS USM (crop factor 1.6)
- etc.

Add 7)
I'm not sure here if the app would be confused by many entries with the same name that only differ in their lens mount

Add 8)
There are a few entries in the database that include the extender

Add 9)
Not sure what the example is about

**Reply by rebio** [reply]
A small update for this discussion and pull request #1649 (Brand based database structure - Discussion #1616 "New structure for database"):

**Migration completed for:**
- actioncams.xml
- mil-canon.xml
- compact-samsung.xml
- slr-canon.xml
- mil-tamron.xml
- slr-tamron.xml

**Issues I saw:**
- lenses in eg. slr-tamron.xml have more than one lens mount. @Sturmflut said "I am not a huge friend of marking the same lens with multiple <mount> entries." I'm not a big fan either. A problem I saw was the _TAMRON SP 90mm  F/2.8 Di VC USD MACRO1:1_ lens, that exists in three entries with Canon EF mount (cropfactor 1.0), Nikon F AF+Canon EF+Sony Alpha mount (cropfactor 1.613) and with Nikon F AF mount (cropfactor 1.0). I personaly would resort the list and separate it by lens mount (See also readme file on sorting [here](https://github.com/rebio/lensfun/blob/brand-based-database-structure/data/brand-db/README.md)). So far entries with more than one lens mount are sorted under "mixed lenses"
- I dug around in EXIF data of my cameras and I discovered that my smartphone does include a manufacturer, but not a lens. I'm not sure how or if you could store calibration data for such a device.

**Questions:**
- Should the lenses of third party manufacturers (eg. Tamron, Sigma, etc.) should be sorted by lens mount? (I would do so)
- Should all mounts be collected in one file? I think this would be a pretty good idea, but I don't know if this would break the database because I don't know how it works and where it expects stuff to be (eg. are all files treated as one giant database, does a file have to include every lens mount it has calibration data for, etc.).

I already had to create a spreadsheet with LibreOffice on my PC that allows me to track all changes done to the files I already migrated. I manually add every commit that is also included in this repo. So every change that was done to the repo since I started the migration in the files listed above is not lost, but already included.
As work progresses, more files will be effected. This leads to more time I have to spend on adding already included commits. Since my time is finite as all of your time as well, I can not use this time for migrating the database.

As already said: It would be nice to have a clear answer on if this migration is in line with the plans for this project or not. I'm fine with both. It's OK if my efforts of the database migration is not welcome. However, in that case I'd rather spend my time with something else that will produce a usable output. So a clear answer would help so that I would not waste more time than necessary.
If the changes are welcome I would request that the branch is made into an official branch in this repo so that it's easier to work with.

EDIT: Ping to all contributers from the first commit page with more than one edit: @tuxfanx @sarunasb @uchrisu

**Reply by gnl21** [reply]
I missed this, and I don't really have very strong feelings about which files we should put the database in, but I think we should be a bit careful about rolling other changes into this kind of reorganisation. I agree that having mil-* and slr-* is probably a bit unhelpful, but I'd be inclined to just do the simple thing and merge them into a single file for manufacturers. I normally end up searching regardless, so larger files that I can search and fewer choices of file work well for me. I don't really want to woodshed this, and I'll get used to whatever gets done here. We don't want to waste too much of @rebio's time maintaining a branch while this is up in the air.

I don't really like the idea of breaking lenses out according to their mounts. If the naming for the lenses on the different mounts is consistent then we should have a single entry for the lens, rather than duplicating the name matching for each mount. (I have a similar feeling about the crop factor, and I've been meaning to propose something for that for a long time. I don't mean to discuss that here, but only mention it because it would be incompatible with grouping the lenses by mount.)

**Comment by sarunasb** [comment]
As far as I understand, XML files are all slurped by the application using liblensfun into one big dataset in memory. So from the library point of view there can be any number of files, with random names or just one big file... The reorganisation then would be for calibrators' eyes mostly or are there other use cases? As a calibrator I very rarely browse/read XML files. `grep` to find what's relevant and when calibration data needs to be added, just append it to the end of appropriate XML (`cat lensfun.xml >>...`). Having sorting by the focal length might be nice, but will it last? — all calibrators will have to remember that when adding any new entries.

It may be just me, but 'm not sure about of how much value is added by doing quite a lot of work on all the XML comments:
```
<!-- Device Name: Canon EOS Rebel T3 -->
<!-- Device Name: Tamron SP 70-300mm f/4-5.6 Di USD -->
```
Wouldn't it be more useful to have them as `<model lang="">` contents? 
Right now they would just add extra noise to, for example, `grep` results (filters can be used, yes, but it's more work).

I understand everyone has own style of working, so it's just my ¢2...

**Comment by rebio** [comment]
Copy from entry #1616:
✘ NOT DONE
- Branch is it right now is already abandoned
- Pull request is closed
- Fork will be deleted soon

**NOTE**
- The branch is abandoned, not the project itself.
- The project how it started is now bound to fail. I simply do not have the time to copy all the commits 1:1 and basically repeat the work of three people for a project with an uncertain end.

**USEFUL/NEEDED CHANGES**
- Working in a branch of a fork does not work. The overhead is too much.
- The sorting has to be done in the live database. Otherwise it's just a commit and version chaos to the max for the person who is sorting.
- Single file for new calibration data: Much of the overhead is from tracking changes in files that are already fully or partially completed. A single file where all the new data can be dumped into would help a lot. Since the database seems to be one giant file for the application anyway it wouldn't matter for a release either.

**COMMENTS**
@gnl21 
- breaking lenses out according to their mounts: I found the table to be easier to sort when the lenses (especially for Canon) are sorted by mounts. Some third party manufacturers (eg. Tamron) have some lenses that only work with the Sony E mount. The problem is of course if you have a generic Tamron lens for a Canon mount and a Nikon mount. Should there be two entries for the same lens? On the other hand: Why not? Maybe they have a slightly different parameters. The bad thing for sure is if we start to have one lens with calibration data for Nikon and Canon (sorted by mount) and one lens that has entries for both mounts in it (sorted in a different section for both mounts).
- crop factor: I would think that a (full frame) lens has different calibration data on a full frame camera and an APS-C camera. I'm not sure how lensfun handles this data. Does lensfun do any checks to see if a full frame lens is used on an APS-C camera and adjusts the parameters? Grouping lenses by crop factor is key in my opinion. The reason simply is that when the database is unsorted as it is right now, updated calibration data can end up in the lens with the wrong crop factor.

@sarunasb 
- XML files are all slurped by the application: If so then it doesn't matter how the database is sorted at all. The entire structure can be optimised towards finding a particular lens in the group of files.
- The reorganisation then would be for calibrators' eyes: In short yes. The target was that a particular device (not only DSLR/MIL but also eg. smartphones) can be found much easier. Eg: I just bought device X. I want to correct the images taken with X using lensfun. I have no idea if there already is data for it. Where do I search/where could I add calibration data? If there is no use in a sorted list, then there is no use in separate files all together. If there was only one file, there would be no question where you would have to search for said device X. If there are separate files, they also should be sorted.
- sorting by the focal length might be nice, but will it last?: There could be a separate file for new entries. (eg. new_data.xml) where all the new entries are thrown into without the need of sorting. Then the entries can be sorted by the willing - or not because it does not seem to matter at all. If there was a "dump" file, calibrators could calibrate and would not have to care about sorting at all. The problem I saw: You have data for a lens with crop factor 1 and 1.6. They are not grouped together, but added to the database when they were added by pull request. Next somebody submits new calibration data. Now you have to go and hunt down the lens and tripple check if it's really the right one.
- XML comments: There are added to have a) seperation between the entires b) to include the manufacturer's name of the device (that might not be the same as the model name). If we use `<model lang="">` what if the manufacturer's name is different from a name that already exists? This is also used to group entries that only appear combined. Compact cameras never appear with a different lens for their camera so there is no reason to not group them. It also makes it easier to search for entries if you add one somewhere in between.
- searching with `grep`: Is grep used by the user or by lensfun? I added the entry as a comment assuming that lensfun would neglect them. If grep is used by the user, it would depend on what they use it for. The idea with the Wikidata item is to have a cross link to Wikidata and therefor Wikimedia Commons and Wikipedia. Many information that is collected over there makes sorting much easier (eg. how many 50mm by Canon exist, are some of them missing?, How do they differ so that they can be sorted in the database with some logic)
